### PR TITLE
More DecisionTree improvements

### PR DESCRIPTION
**Pull request by dellaert**
- Added some depth-first methods: fold, visit, visitWith
- Use template for functions rather than std::functions, which matches against lambdas and function pointers as well.

### CODE REVIEW

File: gtsam/navigation/ImuBias.h [hunk 1]
```diff
@@ -142,10 +142,10 @@ class GTSAM_EXPORT ConstantBias {
     return q - (*this);
   }
   Vector6 GTSAM_DEPRECATED localCoordinates(const ConstantBias& q) {
-    return between(q).vector();
+    return (q - (*this)).vector();
```
Comment (ProfFan) [inline gtsam/navigation/ImuBias.h:145]: Is this inadvertently added?

**Review by ProfFan — APPROVED** [review]
Approving sans one comment.

File: gtsam/navigation/ImuBias.h (hunk 1 above)
Comment (varunagrawal) [reply gtsam/navigation/ImuBias.h:145]: This is fine. It is just addressing the GTSAM deprecation in ImuBias.h.

**Review by varunagrawal — APPROVED** [review]
LGTM

**Pull request merged**